### PR TITLE
Manual SwiGLU implementation

### PR DESCRIPTION
**Pull request by mali-git**
This [PR](https://github.com/Modalities/modalities/pull/154) adds a manual SwiGLU implementation. The original one from xops was imcompatible with activation checkpointing (see issue [#14](https://github.com/Modalities/modalities/issues/14)) 

**General changes:**
* replaces xops swiglu imlementation with custom reimplementation

**Breaking changes:** 
* renaming of `fused_swiglu` to `swiglu` in `ActivationType` (see [here](https://github.com/Modalities/modalities/pull/154/commits/90fb3bd06a407333423cffeab486711e26ef8ddf) for the respective config changes)

### CODE REVIEW

**Review by le1nux — CHANGES_REQUESTED** [review]
The implementation looks correct. 

I added a few ideas on how to improve minor things.
Once addressed we can merge 👍

File: src/modalities/models/gpt2/gpt2_model.py [hunk 1]
```diff
@@ -297,6 +297,31 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         x = self.dropout(x)
         return x
 
+class SwiGLU(nn.Module):
```
[the remark this answers is not in the document]
Comment (mali-git) [reply src/modalities/models/gpt2/gpt2_model.py]: Done in 0b18b12770d84f457779d38ae30a8aec562a8113.

File: src/modalities/models/gpt2/gpt2_model.py [hunk 2]
```diff
@@ -297,6 +297,31 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    def __init__(self, n_embd: int, bias: bool):
+        super().__init__()
+        # Best practice: 4 * n_embd
+        hidden_dim = 256 * ((int(2 * 4 * n_embd / 3) + 256 - 1) // 256)
```
[the remark this answers is not in the document]
Comment (mali-git) [reply src/modalities/models/gpt2/gpt2_model.py]: Done in 0b18b12770d84f457779d38ae30a8aec562a8113.

File: src/modalities/models/coca/multi_modal_decoder.py [hunk 3]
```diff
@@ -33,8 +31,8 @@ def __init__(
 
         if activation == ActivationType.GELU:
             mlp = partial(MLP, in_features=n_embd, hidden_features=ffn_hidden, bias=bias, dropout=dropout)
-        elif activation == ActivationType.FUSED_SWIGLU:
-            mlp = partial(xops.SwiGLU, in_features=n_embd, hidden_features=ffn_hidden, bias=bias)
+        elif activation == ActivationType.SWIGLU:
+            mlp = partial(SwiGLU(n_embd, bias))
```
[the remark this answers is not in the document]
Comment (mali-git) [reply src/modalities/models/coca/multi_modal_decoder.py:35]: @spravil do we need partial here?

File: src/modalities/models/gpt2/gpt2_model.py [hunk 4]
```diff
@@ -297,6 +297,31 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         x = self.dropout(x)
         return x
 
+class SwiGLU(nn.Module):
+    def __init__(self, n_embd: int, bias: bool):
+        super().__init__()
+        # Best practice: 4 * n_embd
```
[the remark this answers is not in the document]
Comment (mali-git) [reply src/modalities/models/gpt2/gpt2_model.py]: I didn't find the reference. @fromm-m, do you remember the reference?

File: src/modalities/models/coca/multi_modal_decoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (spravil) [reply src/modalities/models/coca/multi_modal_decoder.py:35]: We create different constructors depending on the activation function. The constructor is used two times in this particular block and avoids parameter sharing.

But please fix the partial by applying the following change:
```suggestion
            mlp = partial(SwiGLU, n_embd=n_embd, bias=bias)
```

File: src/modalities/models/gpt2/gpt2_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (spravil) [reply src/modalities/models/gpt2/gpt2_model.py]: Maybe https://arxiv.org/pdf/2002.05202 ?

[the remark this answers is not in the document]
Comment (fromm-m) [reply src/modalities/models/gpt2/gpt2_model.py]: "Attention is all you need"
https://arxiv.org/pdf/1706.03762

image

[the remark this answers is not in the document]
Comment (mali-git) [reply src/modalities/models/gpt2/gpt2_model.py]: Done in 58aea7ae56b8fbde209c58dd0adf3f340de208b8.

File: tests/nn/test_mlp.py [hunk 5]
```diff
@@ -8,3 +8,12 @@ def test_mlp_forward():
...
     out = model(dummy_input)
     assert out.shape == (1, 10, 64)
+
+def test_SwiGLU_forward():
+    n_embd = 512
+    bias = True
+    model = SwiGLU(n_embd, bias)
+    input_tensor = torch.randn(1, n_embd)
+    output_tensor = model(input_tensor)
+    assert output_tensor.shape == (1, n_embd)
```
[the remark this answers is not in the document]
Comment (mali-git) [reply tests/nn/test_mlp.py]: Done in 6a65cc974acdd6cfd55a4c671c9903cfe2c32b6b.

**Review by fromm-m — APPROVED** [review]

File: src/modalities/models/model.py [hunk 6]
```diff
@@ -20,6 +20,39 @@ def forward(self, inputs: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
...
+            out_features=hidden_dim,
+            bias=bias,
+        )
+        self.out_proj = nn.Linear(
+            in_features=hidden_dim,
+            out_features=n_embd,
+            bias=bias,
+        )
+
+    def _get_hidden_dim(self, n_embd: int) -> int:
```
Comment (le1nux) [inline src/modalities/models/model.py:46]: we should make this method static

File: tests/nn/test_mlp.py [hunk 7]
```diff
@@ -12,8 +15,30 @@ def test_mlp_forward():
...
-    input_tensor = torch.randn(1, n_embd)
-    output_tensor = model(input_tensor)
-    assert output_tensor.shape == (1, n_embd)
+    mlp = SwiGLU(n_embd, bias)
+
+    hidden_dim = 1536
+    assert mlp._get_hidden_dim(n_embd) == hidden_dim
+    
+    n_embd = 511
+    assert mlp._get_hidden_dim(n_embd) == hidden_dim
```
Comment (le1nux) [inline tests/nn/test_mlp.py:24]: if we make it static: 

```suggestion
    assert SwiGLU._get_hidden_dim(n_embd) == hidden_dim

```

File: src/modalities/models/model.py [hunk 8]
```diff
@@ -20,6 +20,39 @@ def forward(self, inputs: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
...
+        # Best practice: 4 * n_embd (https://arxiv.org/pdf/1706.03762)
+        # Because we add an additional linear layer, we need to adjust the hidden_dim to 2/3 of the original value
```
[the remark this answers is not in the document]
Comment (fromm-m) [reply src/modalities/models/model.py]: addressed in https://github.com/Modalities/modalities/pull/154/commits/4e579b6026b4b0d577892bf68d946af1f8aa545e

File: tests/nn/test_mlp.py (hunk 7 above)
Comment (fromm-m) [reply tests/nn/test_mlp.py]: addressed in https://github.com/Modalities/modalities/pull/154/commits/4e579b6026b4b0d577892bf68d946af1f8aa545e

File: src/modalities/models/model.py (hunk 6 above)
Comment (fromm-m) [reply src/modalities/models/model.py]: addressed in https://github.com/Modalities/modalities/pull/154/commits/4e579b6026b4b0d577892bf68d946af1f8aa545e

**Comment by le1nux** [comment]
fixes #14

**Pull request merged**